third_party/move/move-vm/runtime/src/runtime_type_checks_async.rs

```diff
@@ -0,0 +1,797 @@
+                },
+                Bytecode::Call(idx) => {
+                    return Ok(ExitCode::Call(*idx));
+                },
+                Bytecode::CallGeneric(idx) => {
+                    return Ok(ExitCode::CallGeneric(*idx));
+                },
+                Bytecode::CallClosure(_) => {
+                    return Ok(ExitCode::CallClosure);
+                },
+                Bytecode::Branch(target) => {
+                    frame.pc = *target;
+                    continue;
+                },
+                Bytecode::BrTrue(target) | Bytecode::BrFalse(target) => {
+                    let taken = cursor.consume_cond_br().ok_or_else(|| {
+                        PartialVMError::new_invariant_violation(
+                            "All conditional branches must be recorded",
+                        )
+                    })?;
+
+                    if taken {
+                        frame.pc = *target;
+                    } else {
+                        frame.pc += 1;
+                    }
+                    continue;
+                },
+                Bytecode::StLoc(idx) => {
+                    // Store dummy value - these are not needed for type checks, as we only need to
+                    // know if the value is valid or not.
+                    frame.locals.store_loc(*idx as usize, dummy_local())?;
+                },
+                Bytecode::MoveLoc(idx) => {
+                    frame.locals.move_loc(*idx as usize)?;
+                },
+                // Pack closure is not checked in pre- or post-execution type transition.
+                Bytecode::PackClosure(idx, mask) => {
+                    let function = self.idx_to_loaded_function(frame, *idx)?;
+                    RTTCheck::check_pack_closure_visibility(&frame.function, &function)?;
+                    if RTTCheck::should_perform_checks(&frame.function.function) {
+                        verify_pack_closure(
+                            self.ty_builder,
+                            &mut self.type_stack,
+                            &function,
+                            *mask,
+                        )?;
+                    }
+                },
+                // Pack closure generic is not checked in pre- or post-execution type transition.
+                Bytecode::PackClosureGeneric(idx, mask) => {
+                    let function = self.instantiation_idx_to_loaded_function(frame, *idx)?;
+                    RTTCheck::check_pack_closure_visibility(&frame.function, &function)?;
+                    if RTTCheck::should_perform_checks(&frame.function.function) {
+                        verify_pack_closure(
+                            self.ty_builder,
+                            &mut self.type_stack,
+                            &function,
+                            *mask,
+                        )?;
+                    }
+                },
+                Bytecode::Pop
```

Comment:
can the logic here of matching these to increment the counter or smt be re-used?